### PR TITLE
Mark the replica ERR if its volume.meta is in invalid state

### PR DESCRIPTION
**Pull request by derekbit**
If there is any available replica, mark the replicas ERR if their volume.meta are in invalid state.

https://github.com/longhorn/longhorn/issues/4212

Signed-off-by: Derek Su <derek.su@suse.com>

### CODE REVIEW

**Review by shuo-wu — COMMENTED** [review]
Putting this state check in the revision counter check function looks a little bit weird... 
Will the replica process check the state during the starting stage?

**Comment by derekbit** [comment]
> Putting this state check in the revision counter check function looks a little bit weird... Will the replica process check the state during the starting stage?

Agreed. But I don't know why the original state check is in the revision counter check funciton...

**Comment by derekbit** [comment]
> Will the replica process check the state during the starting stage?

Actually, the check is in `c.backend.GetRevisionCounter(r.Address)`...

**Comment by shuo-wu** [comment]
I may prefer to make this check a separate function, or a part of the replica `Open` function. But due to the complexity of the replica state, more investigation is needed. What do you think?

**Comment by derekbit** [comment]
> 

Yes, I prefer a new function `checkReplicasState` for checking the state before the `checkReplicasRevisionCounter`.

>  due to the complexity of the replica state, more investigation is needed. 

Sure.

File: pkg/controller/control.go [hunk 1]
```diff
@@ -660,6 +697,12 @@ func (c *Controller) Start(addresses ...string) error {
 	// For more details, see the following url:
 	// https://github.com/longhorn/longhorn/issues/1235
 	if !c.isUpgrade {
+		// Check if the replicas are in valid state.
+		// If not in a valid state, mark the replica as ERR.
+		if err := c.checkReplicasState(); err != nil {
+			return err
+		}
```
Comment (shuo-wu) [inline pkg/controller/control.go:704]: I think we can move this as [a part of function `c.Start()`](https://github.com/longhorn/longhorn-engine/blob/7310e09cd6632276f29ae8659a1388b9ee8edbc7/pkg/controller/control.go#L605-L649). 
Then you can eliminate some duplications for `checkReplicasState`

Comment (derekbit) [reply pkg/controller/control.go]: Updated. Thank you.

**Review by shuo-wu — DISMISSED** [review]

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, but just a few feedbacks to resolve.

File: pkg/controller/control.go [hunk 2]
```diff
@@ -577,6 +578,23 @@ func (c *Controller) checkReplicasRevisionCounter() error {
 	return nil
 }
 
+func replicaIsInInvalidState(state string) bool {
```
[the remark this answers is not in the document]
Comment (derekbit) [reply pkg/controller/control.go]: Updated. Thank you.

**Review by innobead — APPROVED** [review]

**Pull request merged**